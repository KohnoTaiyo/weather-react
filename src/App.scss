@use "./styles/variables" as *;

.app {
  max-width: 760px;
  margin: 0 auto;
  padding: 16px;
  min-height: 100vh;

  h1 {
    text-align: center;
    font-size: $titleFontSize;
    margin-bottom: 16px;
  }

  h3 {
    margin-bottom: 8px;
  }

  form {
    margin-bottom: 16px;

    .error {
      color: $redColor;
      font-weight: bold;
      margin-top: 4px;
    }
  }

  .current {
    display: flex;
    align-items: center;
    column-gap: 8px;
    background-color: $primaryColor;
    color: $whiteColor;
    padding: 8px;
    border-radius: 4px;
    margin-bottom: 16px;

    &__name {
      font-size: $largeFontSize;
      font-weight: bold;
    }

    &__temp {
      margin-right: 4px;

      &Main {
        font-size: $titleFontSize;
      }
    }
  }

  .forecast {
    display: flex;
    justify-content: space-between;
    background-color: $primaryColor;
    color: $whiteColor;
    border-radius: 4px;

    &__item {
      width: 100%;
      text-align: center;

      &:not(:last-child) {
        border-right: 1px solid $whiteColor;
      }

      &Date {
        border-bottom: 1px solid $whiteColor;
        padding: 8px;
      }
    }
  }
}
