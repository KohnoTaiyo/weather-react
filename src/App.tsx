import { useEffect, useState } from "react";

import "./App.scss";

import { TextBox } from "./components/TextBox";
import { getForecastWeather } from "./hooks/fetchers";
import { ForecastFewDaysWeatherType } from "./types";

const FORECAST_DAYS_AMOUNT = 5;
const INITIAL_LOCATION = "New York";
const formatDate = (date: string) => date.slice(5).replace("-", "/");

function App() {
  const [text, setText] = useState(INITIAL_LOCATION);
  const [errorMessage, setErrorMessage] = useState("");
  const [forecastWeather, setForecastWeather] = useState<ForecastFewDaysWeatherType | undefined>();
  const [isLoading, setIsLoading] = useState(false);
  const { current, forecast, location } = forecastWeather || {};

  const fetchWeather = async () => {
    setIsLoading(true);
    try {
      const data = await getForecastWeather(text, FORECAST_DAYS_AMOUNT);
      setErrorMessage("");
      setForecastWeather(data);
    } catch (err) {
      if (err instanceof Error) {
        setErrorMessage(err.message);
      }
      setForecastWeather(undefined);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchWeather();
  }, []);

  const onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    try {
      if (!text) throw new Error("Please enter a location.");
      setErrorMessage("");
      await fetchWeather();
    } catch (err) {
      if (err instanceof Error) {
        setErrorMessage(err.message);
      }
    }
  };

  const locationName = [location?.name, location?.region, location?.country].filter(Boolean).join(", ");

  return (
    <main className="app">
      <h1>Weather Forecast</h1>
      <form onSubmit={onSubmit}>
        <TextBox value={text} onChange={setText} />
        {errorMessage && <p className="error">{errorMessage}</p>}
      </form>
      {isLoading ? (
        <p>Loading...</p>
      ) : (
        !errorMessage && (
          <>
            {location && <h3>{`Weather in ${locationName} at ${location.localtime}`}</h3>}
            {current && (
              <div className="current">
                <img
                  src={`https://${current.condition.icon}`}
                  className="current_icon"
                  alt={`${current.condition.text} Image`}
                />
                <p className="current__name">{current.condition.text}</p>
                <p className="current__temp">
                  <span className="current__tempMain">{current.temp_c}</span>Â°C
                </p>
                <p>{`Humidity: ${current.humidity}%`}</p>
                <p>{`Wind: ${current.wind_kph}kph`}</p>
              </div>
            )}
            {forecast && (
              <>
                <h3>5days weather</h3>
                <ul className="forecast">
                  {forecast.forecastday.map((item) => (
                    <li key={item.date} className="forecast__item">
                      <p className="forecast__itemDate">{formatDate(item.date)}</p>
                      <img src={item.day.condition.icon} alt="" />
                      <p className="forecast__itemName">{item.day.condition.text}</p>
                    </li>
                  ))}
                </ul>
              </>
            )}
          </>
        )
      )}
    </main>
  );
}

export default App;
