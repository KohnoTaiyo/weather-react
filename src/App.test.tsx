import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "./App";

const user = userEvent.setup();

describe("App", () => {
  test("input validation", async () => {
    render(<App />);
    const textBox = screen.getByRole("textbox");
    await user.clear(textBox);
    await user.click(screen.getByRole("button", { name: /Search/i }));
    await waitFor(() => {
      screen.getByText(/Please enter a location/i);
    });
  });

  test("renders correctly init", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test("renders correctly rendered", async () => {
    render(<App />);
    await waitFor(() => {
      screen.getByText(/Weather in/i);
      screen.getByText(/5days weather/i);
    });
  });
});
