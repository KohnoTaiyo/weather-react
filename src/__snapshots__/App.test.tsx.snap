// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly init 1`] = `
<div>
  <main
    class="app"
  >
    <h1>
      Weather Forecast
    </h1>
    <form>
      <div
        class="textBox"
      >
        <input
          class="textBox__input"
          placeholder="location"
          required=""
          type="text"
          value="New York"
        />
        <input
          class="textBox__submit"
          tabindex="0"
          type="submit"
          value="Search"
        />
      </div>
    </form>
    <p>
      Loading...
    </p>
  </main>
</div>
`;
